[MRG+2] Timing and training score in GridSearchCV

raghavrv opened the pull request:
Finishing up @eyc88's #7026 

Resolves #6894 and #6895 

**TODO**
- [x] Rebase after merge of #7324 and make all tests pass
- [x] Better tests for training score
- [x] What's new
- [x] Split training and scoring time as `mean|std_fit_time` vs `mean|std_score_time`

~~Have an example at `examples/`?~~ Address in separate PR

@jnothman @amueller @vene @ogrisel (To be reviewed after merge of #7324)

NOTE: Make sure @eyc88 is credited when squashing and merging.

Commit:
Resolved issue #6894 and #6895:
    Now *SearchCV.results_ includes both timing and training scores.

wrote new test (sklearn/model_selection/test_search.py)
and new doctest (sklearn/model_selection/_search.py)
added a few more lines in the docstring of GridSearchCV and RandomizedSearchCV.
Revised code according to suggestions.
Add a few more lines to test_grid_search_results():
    1. check test_rank_score always >= 1
    2. check all regular scores (test/train_mean/std_score) and timing >= 0
    3. check all regular scores <= 1
Note that timing can be greater than 1 in general, and std of regular scores
always <= 1 because the scores are bounded between 0 and 1.

raghavrv commented:
Do we have to get the weighted average for `mean_train_score` across split if i.i.d.?

jnothman commented:
It doesn't really make sense to weight it by test set size. Perhaps by training set size. I'd leave it unweighted.

GaelVaroquaux commented:
> It doesn't really make sense to weight it by test set size. Perhaps by
> training set size. I'd leave it unweighted.

+1.

raghavrv commented:
Yes I was asking for the weighing it by the training set size only... Okay! Let's leave it unweighted...

amueller commented:
are you working on this?

raghavrv commented:
@amueller yes. I was waiting for #7324 to get merged. I'll push in a while...

raghavrv committed:
ENH/FIX timing and training score.

* ENH separate fit / score times
* Make score_time=0 if errored; Ignore warnings in test
* Cleanup docstrings
* ENH Use helper to store the results
* Move fit time computation to else of try...except...else
* DOC readable sample scores
* COSMIT Add a commnent on why time test is >= 0 instead of > 0
  (Windows time.time precision is not accurate enought to be non-zero
   for trivial fits)

raghavrv commented:
I've made the keys as `"mean_time"` and `"std_time"` as the word test/train is irrelevant here...

raghavrv commented:
@amueller @jnothman added some basic tests for training scores... Reviews please...

raghavrv commented:
About the example, do we want one that compares the train and test scores and also plots the training time for each candidate? (Also ping @vene for suggestions)

jnothman commented on sklearn/model_selection/_search.py:
```diff
@@ -746,6 +784,10 @@ class GridSearchCV(BaseSearchCV):
         step, which will always raise the error.
 
+    return_train_score: boolean, default=True
+        If ``'False'``, the results_ attribute will not include training
```
`cv_`

jnothman replied:
space before colon

jnothman commented on doc/whats_new.rst:
```diff
@@ -72,6 +72,17 @@ Model Selection Enhancements and API Changes
 
+  - Training scores and Timing information
+
+    The ``cv_results_`` arrays now include the training scores for each
```
drop "The" "arrays"

"now" -> "also"

"include" -> "includes"

jnothman commented on doc/whats_new.rst:
```diff
@@ -72,6 +72,17 @@ Model Selection Enhancements and API Changes
+    The ``cv_results_`` arrays now include the training scores for each
+    cross-validation split (with keys such as ``'split0_train_score'``), as
+    well as their mean (``'mean_train_score'``) and standard deviation
+    (``'std_train_score'``).
```
"To avoid the cost of evaluating training score, set `return_train_score=False`.

jnothman commented on sklearn/model_selection/_search.py:
```diff
@@ -1006,6 +1059,10 @@ class RandomizedSearchCV(BaseSearchCV):
         FitFailedWarning is raised. This parameter does not affect the refit
         step, which will always raise the error.
 
+    return_train_score: boolean, default=True
```
space before colon

jnothman commented on sklearn/model_selection/_search.py:
```diff
@@ -1006,6 +1059,10 @@ class RandomizedSearchCV(BaseSearchCV):
         step, which will always raise the error.
 
+    return_train_score: boolean, default=True
+        If ``'False'``, the results_ attribute will not include training
```
`cv_`

jnothman commented on sklearn/model_selection/_search.py:
```diff
@@ -1030,16 +1087,25 @@ class RandomizedSearchCV(BaseSearchCV):
 
         NOTE that the key ``'params'`` is used to store a list of parameter
-        settings dict for all the parameter candidates.
+        settings dict for all the parameter candidates.  Besides,
```
Drop "besides ..." This does not need redundant documentation

jnothman commented on sklearn/model_selection/_search.py:
```diff
@@ -806,11 +848,20 @@ class GridSearchCV(BaseSearchCV):
 
         NOTE that the key ``'params'`` is used to store a list of parameter
-        settings dict for all the parameter candidates.
+        settings dict for all the parameter candidates.  Besides,
```
Drop "Besides, ..."

jnothman commented on sklearn/model_selection/_search.py:
```diff
@@ -551,41 +552,78 @@ def _fit(self, X, y, labels, parameter_iterable):
+        time_stds = np.sqrt(np.average((time - time_means[:, np.newaxis]) ** 2,
+                                       axis=1))
 
         cv_results = dict()
```
I should have insisted you keep this local var as `results`!

raghavrv replied:
;)

jnothman commented on sklearn/model_selection/_search.py:
```diff
@@ -551,41 +552,78 @@ def _fit(self, X, y, labels, parameter_iterable):
+                    train_scores[:, split_i])
+            cv_results["mean_train_score"] = train_means
+            cv_results["std_train_score"] = train_stds
+            cv_results["rank_train_score"] = np.asarray(rankdata(-train_means,
```
I don't generally think ranking by training score would be useful / well advised. I'd happily be wrong if someone stated otherwise.

raghavrv replied:
I'm removing it for now... Adding should be easier than deprecating the keys...

jnothman commented on sklearn/model_selection/_search.py:
```diff
@@ -551,41 +552,78 @@ def _fit(self, X, y, labels, parameter_iterable):
+
+        if self.return_train_score:
+            train_means = np.average(train_scores, axis=1)
+            train_stds = np.sqrt(
```
np.std?

raghavrv replied:
Ah yes. Thanks!

jnothman commented on sklearn/model_selection/_search.py:
```diff
@@ -551,41 +552,78 @@ def _fit(self, X, y, labels, parameter_iterable):
+
+        time = np.array(time, dtype=np.float64).reshape(n_candidates, n_splits)
+        time_means = np.average(time, axis=1)
+        time_stds = np.sqrt(np.average((time - time_means[:, np.newaxis]) ** 2,
```
np.std?

jnothman commented on sklearn/model_selection/_search.py:
```diff
@@ -551,41 +552,78 @@ def _fit(self, X, y, labels, parameter_iterable):
+                       weights=weights))
+
+        time = np.array(time, dtype=np.float64).reshape(n_candidates, n_splits)
+        time_means = np.average(time, axis=1)
```
I'm wondering, but not certain, whether it's worth refactoring some of this triplicationl.

raghavrv replied:
It's triplication only if you would be okay with taking the weighted averaging for all the test scores, train scores and times...

raghavrv replied:
Or humm... we could add a helper taking in the weights too...

jnothman commented:
I'm not altogether happy that in `_fit_and_score`, scoring of training data is included within the timing. Should we consider pulling it out?

jnothman commented:
Regarding example, @amueller plotted train score and time in a variant of `plot_rbf_parameters` at #1742.

jnothman commented:
I.e. that was @amueller's attempt at this PR's feature, never merged.

jnothman commented:
Otherwise, re example, I think an example of train scores explicitly introducing over/under-fitting would be interesting, but adding it to the wishlist for a later PR is fine.

raghavrv commented:
> I'm not altogether happy that in _fit_and_score, scoring of training data is included within the timing. Should we consider pulling it out?

I think if we want to then `mean_test_score` would make sense... WDYT?

raghavrv commented:
@ogrisel Now I can bug you here ^_^

jnothman commented:
LGTM.

amueller commented on sklearn/model_selection/tests/test_search.py:
```diff
@@ -642,41 +642,53 @@ def test_grid_search_results():
               dict(kernel=['poly', ], degree=[1, 2])]
     grid_search = GridSearchCV(SVC(), cv=n_splits, iid=False,
-                               param_grid=params)
+                               param_grid=params, return_train_score=True)
```
that's the default, right?

raghavrv replied:
Ah good catch. It should be false by default... Not many would want to see training score...

jnothman replied:
I think it's worth keeping it True by default. Being able to compare training and validation scores is useful and not often costly.

raghavrv replied:
Ok then! Thx!

amueller replied:
yeah I agree with default=True.

amueller replied:
remove the parameter here because it's default?

amueller commented on sklearn/model_selection/tests/test_search.py:
```diff
@@ -689,29 +701,36 @@ def test_random_search_results():
-                                       iid=False, param_distributions=params)
+                                       cv=n_splits, iid=False,
+                                       param_distributions=params,
+                                       return_train_score=True)
```
again, the default...

amueller commented on sklearn/model_selection/tests/test_search.py:
```diff
@@ -730,55 +749,85 @@ def test_search_iid_param():
     random_search = RandomizedSearchCV(SVC(), n_iter=2,
                                        param_distributions={'C': [1, 10]},
-                                       cv=cv)
+                                       cv=cv, return_train_score=True)
```
and again

amueller commented:
there is no test that the training score is computed correctly, is there?

jnothman commented:
> there is no test that the training score is computed correctly, is there?

Wah. Good catch. Indeed, that was the main thing I was waiting for in #7062.

jnothman commented:
> FIX return_train_score should be False by default

But why?

raghavrv commented on sklearn/model_selection/tests/test_search.py:
```diff
@@ -730,55 +749,85 @@ def test_search_iid_param():
+        # For the train scores, we do not take a weighted mean irrespective of
+        # i.i.d. or not
+        assert_almost_equal(train_mean, 1)
+        assert_almost_equal(train_std, 0)
```
These are some really basic tests. I assumed these would suffice @jnothman @amueller

raghavrv commented on sklearn/model_selection/tests/test_search.py:
```diff
@@ -730,55 +749,85 @@ def test_search_iid_param():
         assert_equal(search.cv_results_['param_C'][0], 1)
-        assert_array_almost_equal(cv_scores, [1, 1. / 3.])
+        assert_array_almost_equal(test_cv_scores, [1, 1. / 3.])
+        assert_array_almost_equal(train_cv_scores, [1, 1])
```
And this @jnothman @amueller

jnothman replied:
yeah, fair enough. I might have even checked that off in one reviewing pass ;)

amueller replied:
sorry I don't understand this test.

amueller replied:
oh, never mind, I got it. Hm alright... maybe add a comment?
Having a test for a more standard use would still be nice, but whatever.

raghavrv commented:
> But why?

I assumed it came under an advanced use case / inspection which would not be required by most users? Should I set to `True`?

raghavrv commented:
Done. @amueller Final / One more pass?

amueller commented on sklearn/model_selection/_search.py:
```diff
@@ -746,6 +774,10 @@ class GridSearchCV(BaseSearchCV):
         FitFailedWarning is raised. This parameter does not affect the refit
         step, which will always raise the error.
 
+    return_train_score : boolean, default=True
```
versionadded?

amueller commented on sklearn/model_selection/_search.py:
```diff
@@ -1006,6 +1046,10 @@ class RandomizedSearchCV(BaseSearchCV):
         FitFailedWarning is raised. This parameter does not affect the refit
         step, which will always raise the error.
 
+    return_train_score : boolean, default=True
```
versionadded?

raghavrv replied:
Since `model_selection` is not release yet, do we need a `versionadded`?

amueller replied:
good catch, whoops

amueller commented:
@raghavrv I still find it confusing to explicitly pass the default parameters ;)

raghavrv commented:
I thought I removed them... Give me a moment...

raghavrv commented:
It seems to be removed to me. Could you refresh please?

amueller commented:
hm ok github was messing with me. lgtm on green.

amueller commented:
wait... I was confused... I didn't look at the timing information correctly. I thought there was separate information for training time and prediction time. It doesn't really make sense to me to store them together.

raghavrv commented:
Should I split them into `'mean_training_time'` and `'mean_scoring_time'`... Let's use `training` instead of `train` to distinguish between train set statistic and training process statistic?

And should this scoring time take into account scoring on training set too?

amueller commented:
not sure how meaningful that distinction is. It seems more just like inconsistent naming. And I would just not measure training set scoring time...

raghavrv commented:
`'mean_fit_time'` and `'mean_evaluation_time'`?

amueller commented:
sure why not

raghavrv commented:
Also should there be a `'total_fit_time'` and `'total_evaluation_time'` (for sum across all the splits)?

amueller commented:
-1 on total. We don't need that many attributes

raghavrv commented:
And wait `'mean_validation_time'` instead of `'mean_evaluation_time'`?

amueller commented:
The application I have in mind is searching for models that are fast in predicting, which is something that's often quite useful, and searching for models that are fast to train, which is helpful for automatic model search...

raghavrv commented:
Ahh. Okay

amueller commented:
It's mostly about the time it takes to call ``predict`` so I would rather call it evaluation, which includes both the prediction and the metric computation. I would expect the prediction time to be most of it, though.

vene commented:
I guess the second should be `mean_predict_time` for consistency, but do we even want to measure it?

amueller commented:
@vene see my comment above. And if we call it ``predict`` we should not include the score evaluation, which is actually very hard given the current scorer interface.

vene commented:
Metric computation should be fairly constant, right? But sure, both names are intuitive enough for me. I'm convinced about the application, too, just don't know how to strike thorough the second part on my last comment from mobile.

jnothman commented:
> Metric computation should be fairly constant, right?

Not from metric to metric, but from candidate to candidate, yes.

I propose any of:
* `train_time` and `scoring_time`
* `fit_time` and `scoring_time`
* `fit_time` and `test_time`

"Evaluation" is long and not idiomatic to scikit-learn.

I also suppose `_fit_and_score` should return a dict or a namedtuple to facilitate the proliferation of return arguments.

@amueller, leave this for 0.19? Or at least for 0.18final?

raghavrv commented:
@jnothman Won't there be a confusion between the interpretation of score in `score_time` and `mean_..._score`?

jnothman commented:
English compound nouns are right-headed?

jnothman commented:
I don't find them confusing. Am I the average user?? 😝

amueller commented:
``fit_time`` and ``score_time``?
I'd like to have this in ``0.18-final`` but not if it's polished ;) +1 on returning a dict.

raghavrv commented:
> +1 on returning a dict.

For `cross_val_score` or `_fit_and_score`?

Either way won't dict be wasteful? Should we use a named tuple, one where the base class is generated on the fly based on the number of splits etc?

raghavrv commented:
> I'd like to have this in 0.18-final but not if it's polished 

I think except for splitting the timing and using a named tuple or dict not much needs to be done here...

raghavrv commented:
> I don't find them confusing. Am I the average user??

No you are an outlier in our userbase ;P

jnothman commented:
Conflict.

jnothman commented:
Please rebase. Please call it `score_time` not `scoring_time`. Please integrate these into `cv_results_` and testing.

raghavrv commented:
Done, done and done!

amueller commented:
> For cross_val_score or _fit_and_score?

Definitely not for cross_val_score I'd say.

amueller commented:
flake8

amueller commented on sklearn/model_selection/_search.py:
```diff
@@ -806,6 +854,12 @@ class GridSearchCV(BaseSearchCV):
+            'split1_train_score': [0.95, 0.7, 0.8, 0.8]
+            'mean_train_score'  : [0.93, 0.75, 0.83, 0.9]
+            'std_train_score'   : [0.02, 0.01, 0.03, 0.03],
+            'mean_time'         : [0.00073, 0.00063, 0.00043, 0.00049]
```
shouldn't it be train_time and score_time?

jnothman replied:
fit_time was your edict :)

amueller commented on sklearn/model_selection/_search.py:
```diff
@@ -1030,11 +1090,17 @@ class RandomizedSearchCV(BaseSearchCV):
+            'split1_train_score' : [0.82, 0.5, 0.7],
+            'mean_train_score'   : [0.81, 0.7, 0.7],
+            'std_train_score'    : [0.00073, 0.00063, 0.00043]
+            'mean_time'          : [0.00073, 0.00063, 0.00043]
```
same here

amueller commented on sklearn/model_selection/_validation.py:
```diff
@@ -231,12 +235,16 @@ def _fit_and_score(estimator, X, y, scorer, train, test, verbose,
             raise
         elif isinstance(error_score, numbers.Number):
             test_score = error_score
+            score_time = time.time() - fit_time
```
this seems wrong. ``score_time`` is now negative, right? Test?

jnothman replied:
I think if the score is an error score, score_time is 0.

jnothman replied:
Or NaN

jnothman replied:
I don't think score_time is negative here, but I do think this would be clearer, given the error cases, if you just calculated overall time and `score_time` and subtracted to get `fit_time`.

raghavrv replied:
Yes... it will not be negative...

The way the exception handling is done, overall time cannot be calculated outside the try...except...else... loop as training scores will be included in timing. But I've now set `score_time` to 0 and added a test for timing.

raghavrv commented:
@amueller @jnothman one more pass?

jnothman commented:
I don't think the `_mean_and_std` helper does enough to be helpful, and it remains hard to read which fields are being transformed in which way. Why not let the helper be a closure that makes use of `cv_results`, `n_candidates` and `n_splits` so that the code becomes much more declarative and readable:

```python
if self.return_train_score:
    _store('train_score', train_scores)
_store('test_score', test_scores, weights= test_sample_counts if self.iid else None, splits=True, rank=True)
_store('fit_time', fit_times)
_store('score_time', score_times)
```

Then it's clear from the code exactly what's similar and different across the fields.

raghavrv commented:
Humm... There is a weird failure in appveyor ;(

raghavrv commented:
@jnothman could you take a look and see if it looks okay now?

jnothman commented:
You still have a test failure in Windows.

I'm fine with the helper. To satisfy those of us who don't like local function definitions, btw, you could instead add extra parameters for n_candidates, n_splits and cv_results, but then locally define `_store = functools.partial(_store_cv_results, cv_results=cv_results, n_candidates=n_candidates, n_splits=n_splits)` so that it's not repetitive from line to line. That may improve the code cosmetically; I'm a bit ambivalent.

raghavrv commented:
Yes I'm on it. Thanks for the review :)

I feel it's clearer the way it is... I'll leave it as such unless someone complains...

raghavrv commented:
I think in appveyor's windows, time.time() precision is not high enough to make the fit time > 0. That could be a reason for the test failure :@ I'll try with a larger dataset that takes more time.

amueller commented:
or test >=0?

raghavrv commented:
I just recieved my new laptop with windows installed in it ;) I'll try this out and update soon....

jnothman commented:
oh, you're saying it registered 0. Of course >= 0 is what you want.

On 22 September 2016 at 23:10, Raghav RV <notifications@github.com> wrote:

> I just recieved my new laptop with windows installed in it ;) I'll try
> this out and update soon....
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7325#issuecomment-248899070>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz60HGG_JdHY4QIANK8IJiMoJESHBJks5qsn5fgaJpZM4JyaRM>
> .
>

raghavrv commented:
Apologies for the delay. I got busy configuring my laptop ^_^.

And 

> oh, you're saying it registered 0. Of course >= 0 is what you want.

No I speculate that to be the reason...

I tried running on my windows machine and found it not to be the case... `time.time` is as accurate as it was on linux. I don't know how older versions of windows handle it though...

I still don't have success in installing scikit-learn on window ;(

And if we follow the `>= 0` that would mean we are allowing `fit_time` to be 0 in linux too. Such a test would mask a bug of `fit_time` always being `0`...

In the next couple of hours I'll somehow figure out a way to install scikit-learn on my windows and see if it fails...

raghavrv commented:
I did `>= 0` and the appveyor passes. I am unable to install on windows (I set too small a partition size and it bit me by being unable to install the whopping 8GB for visual studio libraries. I don't understand why mingw requires vs13)

Anyway, @jnothman @amueller WDYT we should do. The current test setup fine?

jnothman commented:
You can use a less trivial fit, then? But I'm okay with this being written
off as "hard to test".

On 24 September 2016 at 09:39, Raghav RV <notifications@github.com> wrote:

> I did >= 0 and the appveyor passes. I am unable to install on windows (I
> set too small a partition size and it bit me by being unable to install the
> whopping 8GB for visual studio libraries. I don't understand why mingw
> requires vs13)
>
> Anyway, @jnothman <https://github.com/jnothman> @amueller
> <https://github.com/amueller> WDYT we should do. The current test setup
> fine?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7325#issuecomment-249326507>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz61B1nXjUChsX_xHAckbXYD-EcUzIks5qtGMWgaJpZM4JyaRM>
> .
>

raghavrv commented:
Finally I succeeded in installing it in windows. (Had to resort to a 32 bit win in virtual box). Now I can confirm that it is the time precision that makes the test fail.

[This section of `timeit`](https://docs.python.org/2/library/timeit.html#timeit.default_timer) explains it clearly. I would say let's leave it as such like you suggested before. The alternative would be to make a `utils.fixes` function `safe_time` that always gives us the highest precision clock. (Which could mean checking platform and choosing between `time.clock`, `time.time`, or the newer `time.perf_counter`)

@amueller @jnothman Final reviews and merge?

jnothman commented on sklearn/model_selection/_search.py:
```diff
@@ -1030,11 +1074,19 @@ class RandomizedSearchCV(BaseSearchCV):
+            'split0_train_score' : [0.8, 0.9, 0.7],
+            'split1_train_score' : [0.82, 0.5, 0.7],
+            'mean_train_score'   : [0.81, 0.7, 0.7],
+            'std_train_score'    : [0.00073, 0.00063, 0.00043],
```
Can we make these numbers more readable? Pretend it's slower.

jnothman reviewed (approved):
Otherwise LGTM

amueller commented:
lgtm.

raghavrv commented:
Thanks everyone for the reviews. @jnothman Does it look fine now? Merge?

raghavrv commented:
@jnothman bump :) Anything more to be done here?

jnothman commented:
Thanks for asking. The last thing to do: mention somewhere that timings are in seconds in `cv_results_`.

raghavrv committed:
Convey that times are in seconds

raghavrv commented:
Done :)

jnothman commented:
And done. Backport please, @amueller...
